### PR TITLE
dbus: add additional tests for dbus methods

### PR DESCRIPTION
**Pull request by dongsupark**
Add additional unit tests for the following DBUS methods.
    
* RestartUnit
* TryRestartUnit
* ReloadUnit
* ReloadOrRestartUnit
* ReloadOrTryRestartUnit
* KillUnit
* ResetFailedUnit
* Reload

### CODE REVIEW

**Comment by dongsupark** [comment]
CI fails only on Debian Stretch, because `ReloadUnit()` doesn't return an expected value. OTOH it works fine on Ubuntu 16.04, as well as on Fedora 24.
Debian Stretch runs with systemd v231, while other distros are with v229. 
Interesting. I'll have a deeper look.

**Comment by dongsupark** [comment]
Updated. PTAL.

Fortunately, the failure I saw earlier today had nothing to with systemd upstream. It was just because `/bin/kill` didn't exist in the Debian Docker image. Fixed by calling `systemctl kill` instead. Besides I also fixed other minor issues to get it working on Ubuntu, Debian, and Fedora.

**Comment by lucab** [comment]
ok to test

File: dbus/methods_test.go [hunk 1]
```diff
@@ -130,14 +135,209 @@ func TestStartStopUnit(t *testing.T) {
...
+		t.Fatalf("Test unit not found in list")
+	} else if unit.ActiveState != "active" {
+		t.Fatalf("Test unit not active")
+	}
+
+	// Reload the unit
+	reschan = make(chan string)
+
+	// NOTE: As ReloadUnit() sometimes returns "failed", let's
+	// retry 3 times until it's returning the correct status.
```
Comment (dongsupark) [inline dbus/methods_test.go]: @lucab I've seen the race only on Fedora 24 with systemd v229. OTOH it seems to work fine on Debian/Ubuntu. The race doesn't always happen. Not sure if it's an upstream issue or a go-systemd issue.

File: dbus/methods_test.go [hunk 2]
```diff
@@ -523,6 +732,107 @@ func TestStartStopTransientScope(t *testing.T) {
...
+	maxAttempts := 3
+	for i := 0; i < maxAttempts; i++ {
+		conn.KillUnit(target, int32(syscall.SIGTERM))
+
+		units, err := conn.ListUnits()
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		unit := getUnitStatus(units, target)
```
Comment (dongsupark) [inline dbus/methods_test.go]: @lucab Hmm, that sounds like a good idea. Will try it. I also think such a polling is ugly.

**Comment by dongsupark** [comment]
Updated. PTAL.

* In `TestReloadUnit()` and `TestKillUnit()`, replaced arbitrary polling with a pair of unit subscription and waiting for the events. As `dbus.Conn.Subscribe()` registers a signal `PropertiesChanged` to `org.freedesktop.DBus.Properties`, caller only needs to wait for an event channel to check the returned unit status.

* Split a part of `TestReloadUnit` into `TestReloadOrRestartUnit` for better readability.

**Comment by lucab** [comment]
This LGTM, thanks. @jonboulle do you want to have a pass on this?

**Comment by lucab** [comment]
No concerns from @jonboulle, merging.

**Comment by dongsupark** [comment]
@lucab thanks!

**Comment by lucab** [comment]
@dongsupark thank you for taking care of the testsuite!

**Pull request merged**